File: kotlin/clock/src/main/kotlin/Clock.kt
class Clock(var h: Int, var m: Int) {

    companion object {
        const val DAY_MINUTES = 24 * 60
    }

    init {
        normalize()
    }


    fun subtract(minutes: Int) {
        m -= minutes
        normalize()
    }

    fun add(minutes: Int) {
        m += minutes
        normalize()
    }

    private fun normalize() {
        val mins = h * 60 + m
        val normalized = ((mins % DAY_MINUTES) + DAY_MINUTES) % DAY_MINUTES
        h = normalized / 60
        m = normalized % 60
    }

    override fun hashCode(): Int {
        var result = h
        result = 31 * result + m
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (other is Clock) {
            return h == other.h && m == other.m
        }

        return false
    }

    override fun toString(): String {
        return "${h.toString().padStart(2, '0')}:${m.toString().padStart(2, '0')}"
    }
}

File: csharp/clock/Clock.cs
public class Clock
{
    private static readonly int DAY_MINUTES = 24 * 60;

    private int hours;
    private int minutes;

    public Clock(int hours, int minutes)
    {
        this.hours = hours;
        this.minutes = minutes;
        Normalize();
    }

    public Clock Add(int minutesToAdd)
    {
        minutes += minutesToAdd;
        Normalize();
        return this;
    }

    public Clock Subtract(int minutesToSubtract)
    {
        minutes -= minutesToSubtract;
        Normalize();
        return this;
    }

    private void Normalize()
    {
        var mins = hours * 60 + minutes;
        var normalized = ((mins % DAY_MINUTES) + DAY_MINUTES) % DAY_MINUTES;
        hours = normalized / 60;
        minutes = normalized % 60;
    }

    public override bool Equals(object? obj) => obj is Clock other && hours == other.hours && minutes == other.minutes;

    public override int GetHashCode() => HashCode.Combine(hours, minutes);

    public override string ToString() => $"{hours:00}:{minutes:00}";
}
